RealmSwift/Object.swift

```diff
@@ -273,12 +273,15 @@ open class Object: RLMObjectBase, ThreadConfined {
      Returns whether two Realm objects are equal.
 
      Objects are considered equal if and only if they are both managed by the same Realm and point to the same
-     underlying object in the database.
+     underlying object in the database. Object comparison only takes place if the object type has a primary key.
 
      - parameter object: The object to compare the receiver to.
      */
-    open override func isEqual(_ object: Any?) -> Bool {
-        return RLMObjectBaseAreEqual(self as RLMObjectBase?, object as? RLMObjectBase)
+    open override func isEqual(to object: Any?) -> Bool {
+        if let object = object as? Object {
+            return RLMObjectBaseAreEqual(self, object)
```

Comment:
Shouldn't this only be calling `RLMObjectBaseAreEqual` if the object has a primary key?

The logic in `-[RLMObjectBase isEqual:]` looks like so:
```objc
- (BOOL)isEqual:(id)object {
    if (RLMObjectBase *other = RLMDynamicCast<RLMObjectBase>(object)) {
        if (_objectSchema.primaryKeyProperty) {
            return RLMObjectBaseAreEqual(self, other);
        }
    }
    return [super isEqual:object];
}
```